### PR TITLE
[codex] Add KA lifecycle CLI devnet suite

### PR DESCRIPTION
**Pull request by Jurij89**
## Summary

- Adds a dedicated devnet suite for the lifecycle-native Knowledge Asset CLI.
- Covers live `dkg ka` one-shot create/write/finalize/share, sub-graph scoping, async SWM share jobs, async VM publish enqueue, `publisher publish-async` alias behavior, pull/query/discard, and removed direct/legacy route guards.
- Registers the suite in the devnet manifest, root scripts, workspace config, and lockfile so it participates in PR-coverage sweeps instead of living as an orphan test.

## Related

- Follow-up to #1410

## Files changed

| File | What |
|------|------|
| `devnet/ka-lifecycle-cli/automated.test.ts` | New live devnet coverage for the KA lifecycle CLI and removed legacy surfaces |
| `devnet/ka-lifecycle-cli/package.json` | Declares the devnet suite package |
| `devnet/ka-lifecycle-cli/vitest.config.ts` | Adds the suite Vitest config with Windows-safe include normalization |
| `devnet/suites.json` | Registers the suite in `prCoverage` and `all` |
| `package.json` | Adds `test:devnet:ka-lifecycle-cli` |
| `pnpm-workspace.yaml` | Adds the suite workspace package |
| `pnpm-lock.yaml` | Adds the suite importer |

## Test plan

- [x] `pnpm install --lockfile-only --ignore-scripts`
- [x] Direct Node manifest invariant check for `devnet/suites.json`, `pnpm-workspace.yaml`, and root `test:devnet:*` script coverage
- [x] `pnpm exec tsx -e "void import('./devnet/ka-lifecycle-cli/vitest.config.ts').then(() => console.log('config import ok'))"`
- [x] `pnpm exec vitest run --config devnet/ka-lifecycle-cli/vitest.config.ts --reporter dot` discovers the suite and fails only at the expected no-live-devnet precondition in this local worktree
- [x] `git diff --cached --check`
- [ ] Run `pnpm run build && ./scripts/devnet.sh start 6 && pnpm test:devnet:ka-lifecycle-cli` on a live devnet

### CODE REVIEW

File: devnet/ka-lifecycle-cli/automated.test.ts [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+  return id;
+}
+
+function parseShareJobId(stdout: string): string {
+  const jobId = /^\s*Job ID:\s+(.+)$/m.exec(stdout)?.[1]?.trim();
+  if (!jobId) throw new Error(`could not parse share job id from:\n${stdout}`);
+  return jobId;
+}
+
+function lastJsonBlock(stdout: string): Record<string, unknown> {
```
Comment (otReviewAgent) [inline devnet/ka-lifecycle-cli/automated.test.ts]: 🟡 Issue: Centralize the CLI JSON-output parser instead of copying it into another suite

**What's wrong**
This is generic devnet harness behavior, not KA-lifecycle-specific logic. Adding another local copy makes each suite responsible for preserving the same CLI-output quirks and encourages more cast-heavy parsing at call sites as additional CLI devnet suites are added.

**Example**
`pr1388-okf-integration` already parses the final JSON object from CLI stdout with the same scan/backoff loop; this suite now carries its own copy and casts each result shape locally.

**Suggested direction**
Promote this into the shared devnet harness, preferably with a useful failure label and either a generic return type or small command-specific wrappers where callers need structured fields.

**For Agents**
Move the scan-for-last-JSON-object behavior into `devnet/_bootstrap/harness.ts` as a shared CLI stdout parser, preserve current parsing behavior, replace the local helpers in this suite and the OKF suite, and run typecheck or the affected devnet vitest configs if available.

File: devnet/ka-lifecycle-cli/automated.test.ts [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+        contextGraphId,
+        '--sub-graph-name',
+        subGraphName,
+        '--json',
+      ],
+      60_000,
+    );
+    expectCliOk(publish, 'ka publish-async');
+    const body = lastJsonBlock(publish.stdout);
+    expect(body.jobId, `publish-async response: ${publish.stdout}`).toBeTruthy();
```
Comment (otReviewAgent) [inline devnet/ka-lifecycle-cli/automated.test.ts:402]: 🟡 Issue: Async publish coverage stops at enqueue acceptance

**What's wrong**
The changed devnet coverage is meant to verify `ka publish-async` and the `publisher publish-async` alias as operational VM publish paths, including sub-graph scoping. A `jobId` plus `accepted` only proves the daemon accepted the request; it does not verify the async worker can publish the named KA or that the sub-graph scope is honored after enqueue.

**Example**
A regression could enqueue `dkg ka publish-async --sub-graph-name sg` successfully but then have the publisher job fail, ignore the subGraphName during execution, or never materialize the KA in VM. This test would still pass because it stops at `accepted`.

**Suggested direction**
Follow the async publish job through completion and assert the published outcome, not just the immediate enqueue response.

**For Agents**
In `devnet/ka-lifecycle-cli/automated.test.ts`, after parsing the async publish jobId, poll `dkg publisher job <jobId>` or the publisher job API until a terminal state. Fail on `failed`, assert `finalized` or the suite's accepted terminal state, and for the sub-graph case prove the resulting VM/sub-graph content is visible or that the finalized job preserved the subGraphName/intent for the named KA. Apply the same pattern to the `publisher publish-async` alias test or share a helper.

File: devnet/ka-lifecycle-cli/automated.test.ts [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+      // keep scanning
+    }
+  }
+  return JSON.parse(stdout.trim()) as Record<string, unknown>;
+}
+
+function writeNtFixture(name: string, value: string): { filePath: string; subject: string } {
+  const dir = join(import.meta.dirname, 'turns');
+  mkdirSync(dir, { recursive: true });
+  const subject = `urn:test:pr1410:${name}:${Date.now()}:${++fileCounter}`;
```
Comment (otReviewAgent) [inline devnet/ka-lifecycle-cli/automated.test.ts]: 🔵 Nit: Remove the stale PR-specific prefix from generated suite artifacts

**Why it matters**
This suite is being added as a reusable named devnet suite rather than a PR-numbered one, so stale PR labels in generated IDs make devnet logs and persisted artifacts misleading.

**Suggestion**
Use the suite name as the stable prefix, for example `ka-lifecycle-cli` / `kaLifecycleCli`, instead of `pr1410` in the subject, context-graph slug, and subgraph name.

File: devnet/ka-lifecycle-cli/automated.test.ts (hunk 1 above)
Comment (Jurij89) [reply devnet/ka-lifecycle-cli/automated.test.ts]: Addressed in d8e3904e: moved the final-JSON stdout parser into devnet/_bootstrap/harness.ts as parseLastJsonBlock, replaced the local KA and OKF copies, and verified both affected devnet configs import. The OKF config also now normalizes the Windows include path so the affected suite is discoverable locally.

File: devnet/ka-lifecycle-cli/automated.test.ts (hunk 2 above)
Comment (Jurij89) [reply devnet/ka-lifecycle-cli/automated.test.ts:402]: Addressed in d8e3904e: the KA devnet suite now parses the async publish job id, polls publisher job <id> --payload until finalized, fails on failed, verifies the preserved knowledge-asset publish request including context graph/name/subgraph/intent key, and queries VM visibility. The same helper is used by both dkg ka publish-async and the publisher publish-async alias test.

File: devnet/ka-lifecycle-cli/automated.test.ts (hunk 3 above)
Comment (Jurij89) [reply devnet/ka-lifecycle-cli/automated.test.ts]: Addressed in d8e3904e: replaced the stale pr1410-generated subject, context graph slug, and subgraph prefixes with durable ka-lifecycle-cli naming.

File: devnet/ka-lifecycle-cli/automated.test.ts (hunk 2 above)
Comment (otReviewAgent) [reply devnet/ka-lifecycle-cli/automated.test.ts:402]: 🔴 Bug: The pull-from/discard coverage can pass without those operations changing storage

**What's wrong**
The test already has the target subject in WM before it calls `pull-from`, and it never proves that WM was cleared or replaced. That makes the later query a weak assertion: stale draft data can satisfy it. The final discard assertion has the same shape because it trusts the JSON flag rather than verifying the graph state.

**Example**
A regression where `ka pull-from --layer swm --on-conflict replace` returns 200 but never replaces the WM draft would still pass, because the subject written at lines 306-313 can still be returned by `ka query` at lines 360-365. Similarly, `discard` can return `{ discarded: true }` while leaving quads behind and this test would not notice.

**Suggested direction**
Add pre/post state assertions so the test proves pull-from seeds from SWM and discard removes the reopened WM draft, instead of only proving the CLI command exited successfully.

**For Agents**
In `devnet/ka-lifecycle-cli/automated.test.ts`, make the pull-from and discard effects observable. Clear or otherwise invalidate the WM draft before pull-from, assert it is absent, then pull from SWM and assert the subject returns. After discard, query the KA again and assert the subject/quads are gone or the expected empty/error state is returned.

Comment (otReviewAgent) [reply devnet/ka-lifecycle-cli/automated.test.ts:402]: 🟡 Issue: The `--share` test does not prove it avoids VM publishing

**What's wrong**
The suite’s stated behavior includes “without VM publishing,” but the test never checks VM state before calling `ka publish-async`. It only checks stdout and SWM visibility, so an unintended VM publish during create/share would not be caught reliably.

**Example**
If `ka create --share` accidentally also enqueued or performed a VM publish, this test could still pass: SWM would be visible, stdout might omit `UAL`, and the later `publish-async`/VM visibility checks would still succeed against an already-published subject.

**Suggested direction**
Add a pre-publish assertion that the unique subject is not visible in verifiable memory, or validate the publisher job surface has no job for this KA before the explicit publish command.

**Confidence note**
There may be lower-level unit coverage that checks the CLI request body, but this devnet suite explicitly documents network-observable behavior and does not verify the network-visible absence of VM publication.

**For Agents**
In the first devnet test, after `assertSharedSubject` and before `ka publish-async`, add a network-visible negative check for the unique subject in VM and/or assert no publisher job exists for that KA name. Keep the existing post-publish VM assertion to prove the explicit publish path still works.

File: devnet/ka-lifecycle-cli/automated.test.ts [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+      subGraphName?: string;
+      intentKey?: string;
+    };
+  };
+  finalization?: Record<string, unknown>;
+  failure?: unknown;
+  error?: unknown;
+}
+
+const suite: Partial<Suite> = {};
```
Comment (otReviewAgent) [inline devnet/ka-lifecycle-cli/automated.test.ts]: 🟡 Issue: Avoid Partial suite state plus casts for initialized test context

**What's wrong**
The new suite makes its core test context optional and then bypasses that optionality at every use site. That weakens the type boundary around the setup contract and makes future changes harder to review because TypeScript no longer proves that `beforeAll` populated every required field.

**Example**
If a future edit adds a required field to `Suite` but forgets to assign it in `beforeAll`, the `suite as Suite` casts still compile and the failure moves to runtime as an undefined value in whichever test first uses it.

**Suggested direction**
Model the real invariant directly: the suite context is either fully initialized by `beforeAll` or the suite fails. A definite `let suite: Suite` or a tiny checked `getSuite()` helper removes the optional fields and the repeated casts.

**For Agents**
In `devnet/ka-lifecycle-cli/automated.test.ts`, replace the partial object with `let suite: Suite;`. Build `node`, `contextGraphId`, and `subGraphName` locally in `beforeAll`, then assign `suite = { node, contextGraphId, subGraphName }` once. Remove the `as Suite` casts. Preserve the current setup behavior and run the KA lifecycle devnet suite or at least the project typecheck if a devnet is unavailable.

File: devnet/ka-lifecycle-cli/automated.test.ts [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+  const filePath = join(dir, `${name}-${fileCounter}.nt`);
+  writeFileSync(
+    filePath,
+    `<${subject}> <${PREDICATE}> "${value}" .\n`,
+    'utf8',
+  );
+  return { filePath, subject };
+}
+
+async function assertSharedSubject(
```
Comment (otReviewAgent) [inline devnet/ka-lifecycle-cli/automated.test.ts]: 🟡 Issue: Collapse the duplicated subject visibility pollers

**What's wrong**
The suite adds two near-identical polling helpers. This is small today, but it puts the query shape, subgraph scoping, and literal comparison policy in two places, so the next visibility assertion or query-shape change has to be kept in sync manually.

**Example**
Both helpers build `SELECT ?o WHERE { GRAPH ?g { <subject> <predicate> ?o } }`, pass `{ contextGraphId, view, ...(subGraphName ? { subGraphName } : {}) }`, return rows only when `lexical(row.o) === value`, and then assert `rows.length > 0`.

**Suggested direction**
Extract one parameterized polling helper for subject visibility, with `view: 'shared-working-memory' | 'verifiable-memory'` and timing options. That deletes one whole branch of duplicated orchestration while keeping the test behavior unchanged.

**For Agents**
In `devnet/ka-lifecycle-cli/automated.test.ts`, introduce a single `assertSubjectVisible` helper that takes the view, timeout, interval, and optional subgraph. Keep small wrappers only if they improve call-site readability. Preserve the current SWM and VM timeout values and existing assertions.

File: devnet/ka-lifecycle-cli/automated.test.ts (hunk 2 above)
Comment (Jurij89) [reply devnet/ka-lifecycle-cli/automated.test.ts:402]: Addressed the new coverage follow-ups in c192a26d2: the one-shot --share test now asserts the unique subject is absent from VM and that the publisher job surface has no VM-publish job for the KA before the explicit publish command. The stepwise test now discards the WM draft before pull-from, verifies it is absent, pulls from SWM, verifies it returns, then discards again and verifies WM no longer contains the subject.

File: devnet/ka-lifecycle-cli/automated.test.ts (hunk 4 above)
Comment (Jurij89) [reply devnet/ka-lifecycle-cli/automated.test.ts]: Addressed in c192a26d2: replaced the Partial<Suite> object with definite let suite: Suite assignment after beforeAll completes, and removed the repeated as Suite casts.

File: devnet/ka-lifecycle-cli/automated.test.ts (hunk 5 above)
Comment (Jurij89) [reply devnet/ka-lifecycle-cli/automated.test.ts]: Addressed in c192a26d2: collapsed the duplicated SWM/VM polling logic into a parameterized assertSubjectVisible helper while preserving the existing per-view timeouts and small readability wrappers.

File: devnet/ka-lifecycle-cli/automated.test.ts (hunk 2 above)
Comment (otReviewAgent) [reply devnet/ka-lifecycle-cli/automated.test.ts:402]: 🔴 Bug: The no-publish assertion ignores jobs with the wrong subgraph

**What's wrong**
The test claims to prove `ka create --share` does not enqueue VM publishing, but the matcher filters out same-KA jobs unless the subgraph also matches. That creates false confidence for the exact regression where the publish request exists but loses subgraph scoping.

**Example**
If `ka create --share` regresses and enqueues a `knowledge-asset-vm-publish` job for the same `contextGraphId` and `name` but drops `subGraphName`, this predicate returns `false`; the negative assertion passes even though a VM publish was enqueued.

**Suggested direction**
Use a separate matcher for the negative assertion that matches by `jobType`, `contextGraphId`, and `name` regardless of `subGraphName`, or explicitly fail with a clearer message when a same-KA job has a missing/wrong subgraph.

**For Agents**
Update the negative assertion path in `devnet/ka-lifecycle-cli/automated.test.ts` so any `knowledge-asset-vm-publish` job for the same context graph and KA name fails the `create --share does not VM publish` check, while keeping exact subgraph checks for positive publish assertions.

File: devnet/_bootstrap/harness.ts [hunk 6]
```diff
@@ -94,6 +94,31 @@ export interface CliResult {
...
+/**
+ * Extract the final JSON object a CLI command prints after human/log lines.
+ * Many devnet suites use JSON mode on commands that still emit progress text;
+ * scanning from the last object start keeps those assertions consistent.
+ */
+export function parseLastJsonBlock<T extends Record<string, unknown> = Record<string, unknown>>(
+  stdout: string,
+  label = 'CLI stdout',
+): T {
+  for (let i = stdout.lastIndexOf('\n{'); i >= 0; i = stdout.lastIndexOf('\n{', i - 1)) {
```
Comment (otReviewAgent) [inline devnet/_bootstrap/harness.ts]: 🔴 Bug: Malformed JSON at the beginning of stdout can hang the parser

**What's wrong**
The scan can revisit the same match forever when the only candidate starts at byte 0 and does not parse. That turns bad CLI output into a synchronous infinite loop, which can freeze devnet test execution instead of producing an actionable failure.

**Example**
parseLastJsonBlock('\n{not-json', 'x') never reaches the fallback or error path: it finds index 0, JSON.parse fails, then lastIndexOf('\n{', -1) returns 0 again forever. Expected behavior is to stop scanning and throw the "no JSON object" error.

**Suggested direction**
Make the scan termination explicit when the matched '\n{' is at index 0, then fall through to the existing whole-stdout parse/error path.

**For Agents**
In devnet/_bootstrap/harness.ts, update parseLastJsonBlock so the search moves strictly earlier and terminates at index 0, for example by breaking after a failed parse at i === 0 or by using a nextIndex variable that becomes -1. Preserve successful parsing of whole-stdout JSON and log-prefixed final JSON. Add a small unit-level check or local assertion for stdout beginning with '\n{' but not valid JSON.

The shared JSON parser lacks coverage for a hanging scan path

**What's wrong**
This helper now underpins multiple devnet validation suites, but there is no direct test for the parse-failure path at offset 0. A malformed CLI output that starts with `\n{` can hang the suite instead of failing with the intended diagnostic.

**Example**
A focused unit case like `parseLastJsonBlock('\n{not-json')` would never reach the fallback or throw; it keeps retrying offset 0.

**Suggested direction**
Add a regression test around the leading-newline object-start edge case and adjust the loop to break or move earlier when `i` is 0.

**For Agents**
Add small unit coverage for `parseLastJsonBlock` in the harness tests, including malformed leading `\n{` output and a normal log-plus-final-JSON case. Fix the scan so the `i === 0` failure path terminates before relying on the helper in more suites.

File: devnet/ka-lifecycle-cli/automated.test.ts (hunk 2 above)
Comment (Jurij89) [reply devnet/ka-lifecycle-cli/automated.test.ts:402]: Addressed the same-KA wrong-subgraph gap in bd6e3c7ad: the no-publish negative assertion now fails on any knowledge-asset-vm-publish job for the same context graph and KA name, regardless of subGraphName. The positive explicit publish path still checks the exact subgraph on the finalized job.

File: devnet/_bootstrap/harness.ts (hunk 6 above)
Comment (Jurij89) [reply devnet/_bootstrap/harness.ts]: Addressed in bd6e3c7ad: parseLastJsonBlock now terminates after a failed offset-0 candidate instead of revisiting the same match, then falls through to the existing whole-stdout parse/error path. Added devnet/_bootstrap/harness-json.test.ts covering log-plus-final-JSON and the malformed leading newline case; the no-devnet bootstrap manifest config now runs that test. Verified with pnpm exec vitest run --config devnet/_bootstrap/vitest.manifest.config.ts --reporter dot.

File: devnet/ka-lifecycle-cli/automated.test.ts (hunk 1 above)
Comment (otReviewAgent) [reply devnet/ka-lifecycle-cli/automated.test.ts]: 🟡 Issue: Move the lifecycle mini-harness out of the scenario file

**What's wrong**
The new suite is not over 1k lines, but it already mixes scenario coverage with a substantial local harness. That makes the workflow harder to scan, weakens reuse, and works against the shared devnet harness direction already present in the codebase. The cleaner move is to separate orchestration helpers from the actual scenario assertions before this grows further.

**Example**
A reader trying to understand the first scenario has to scan ~260 lines of local mini-harness first, including `expectCliOk`, `parseJobId`, fixture generation, subject visibility polling, publisher-job polling, and query normalization. The next KA lifecycle devnet suite is likely to copy the same block instead of reusing it.

**Suggested direction**
Keep `automated.test.ts` as the behavior narrative and move reusable command helpers, fixture creation, publisher-job polling, and WM/SWM/VM assertion helpers behind a small lifecycle driver. That would delete a lot of local scaffolding from the test file without changing behavior.

**For Agents**
Extract the KA lifecycle command/test driver pieces from `devnet/ka-lifecycle-cli/automated.test.ts` into a focused local module such as `devnet/ka-lifecycle-cli/harness.ts`, or into `_bootstrap` if other suites will use them. Preserve the current CLI calls and assertions; the scenario file should mostly read as the four workflows under test.

File: devnet/ka-lifecycle-cli/automated.test.ts (hunk 4 above)
Comment (otReviewAgent) [reply devnet/ka-lifecycle-cli/automated.test.ts]: 🟡 Issue: Replace inline cast-heavy JSON handling with typed response parsers

**What's wrong**
The suite centralizes raw JSON extraction, but the actual CLI wire shapes are still represented as loose optional records and inline casts. That hides the invariant each command is expected to satisfy, makes failures less local, and makes future maintenance harder because readers cannot tell which fields are required versus tolerated noise.

**Example**
If `ka share-jobs --json` returns `{ items: [...] }` or `{ jobs: { ... } }`, line 458 quietly turns that into an empty typed array and the test later fails as “job missing”. The structural problem is that the CLI response boundary is not modeled in one place, so shape drift is diagnosed downstream instead of at parse time.

**Suggested direction**
Push the JSON boundary into a few explicit parsers/type guards rather than spreading optional bags and casts through the scenarios. `parseLastJsonBlock` can remain the raw extraction primitive, but each command shape should be checked once before the workflow uses it.

**For Agents**
Look at `PublisherJobView`, `KnowledgeAssetQueryResult`, and all `parseLastJsonBlock` call sites in `devnet/ka-lifecycle-cli/automated.test.ts`. Preserve the current CLI behavior, but add small shape-specific parsers or type guards like `parseAcceptedJob`, `parsePublisherJobView`, and `parseShareJobsList` that validate the minimal expected fields and return typed values with clear errors.

File: devnet/ka-lifecycle-cli/automated.test.ts [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+  });
+
+  it('deleted direct publish and legacy SWM routes remain unavailable', async () => {
+    const { node, contextGraphId } = suite;
+    const { filePath } = writeNtFixture('legacy-top-level', 'legacy route guard');
+
+    const legacyCli = await runDkgCli(node, ['publish', contextGraphId, '--file', filePath], 30_000);
+    expect(legacyCli.code, `top-level publish unexpectedly succeeded:\n${legacyCli.stdout}\n${legacyCli.stderr}`)
+      .not.toBe(0);
+    expect(`${legacyCli.stdout}\n${legacyCli.stderr}`).toMatch(/unknown command|error/i);
```
Comment (otReviewAgent) [inline devnet/ka-lifecycle-cli/automated.test.ts]: 🔴 Bug: The removed top-level publish guard can pass even if the command still exists

**What's wrong**
This gives false confidence for one of the behaviors the new devnet suite is supposed to verify: that the legacy direct CLI publish surface is gone. A command that still exists but fails later would satisfy the current assertion.

**Example**
If `dkg publish <cg> --file <path>` were accidentally re-registered but failed by calling the removed API route and printing `Error: POST /api/knowledge-assets/publish has been removed`, this assertion would still pass because the exit code is non-zero and the output matches `/error/i`. The legacy CLI surface would be present, but the test would report green.

**Suggested direction**
Assert the specific unavailable-command behavior instead of allowing any generic error from a still-registered command.

**For Agents**
In `devnet/ka-lifecycle-cli/automated.test.ts`, tighten the legacy CLI assertion so it proves the command surface is absent. Prefer matching Commander’s unknown-command text for `publish`, or separately assert that `dkg --help` does not list a top-level `publish` command while keeping the route-removal assertions for the HTTP APIs.

Comment (Jurij89) [reply devnet/ka-lifecycle-cli/automated.test.ts]: Addressed in cdd6bf002: the removed top-level publish guard now requires Commander unknown-command output for publish and also checks dkg --help does not list a top-level publish command. The HTTP removed-route assertions remain separate. Verified with KA suite config import, devnet-suite discovery to the expected no-devnet precondition, and git diff --check.

File: devnet/_bootstrap/harness.ts [hunk 8]
```diff
@@ -94,6 +94,31 @@ export interface CliResult {
   stderr: string;
 }
 
+/**
+ * Extract the final JSON object a CLI command prints after human/log lines.
+ * Many devnet suites use JSON mode on commands that still emit progress text;
+ * scanning from the last object start keeps those assertions consistent.
+ */
+export function parseLastJsonBlock<T extends Record<string, unknown> = Record<string, unknown>>(
```
Comment (otReviewAgent) [inline devnet/_bootstrap/harness.ts:102]: 🟡 Issue: Avoid making stdout scraping the shared JSON boundary

**What's wrong**
The new helper centralizes duplicated code, but it also normalizes a weak contract: commands in JSON mode may still mix human logs and machine data on stdout, and every caller must remember to scrape the last object. That spreads an incidental formatting dependency through the harness instead of making the CLI/test boundary explicit.

**Example**
A future CLI command can keep printing progress lines to stdout in --json mode and still look acceptable to tests by adding another parseLastJsonBlock call. That makes the machine-output boundary less clear instead of forcing one canonical shape: either JSON-only stdout, or a harness wrapper that explicitly owns the mixed-output compatibility behavior.

**Suggested direction**
Do not export a raw generic parser as the canonical abstraction. Push the boundary up to command execution with runDkgCliJson<T>(), or push it down into the CLI so --json stdout is clean and tests can use JSON.parse directly.

**Confidence note**
This is a structural concern: the mixed stdout behavior may be pre-existing, but this PR promotes the workaround from one suite-local helper into a shared harness API and adds new call sites.

**For Agents**
Look at devnet/_bootstrap/harness.ts and the parseLastJsonBlock call sites. Preserve current test behavior, but prefer either fixing --json commands to emit clean JSON stdout or introducing a higher-level runDkgCliJson<T>() wrapper that runs the command, parses compatibility output privately, and lets tests assert typed command responses without directly depending on stdout scraping.

File: devnet/ka-lifecycle-cli/automated.test.ts [hunk 9]
```diff
@@ -0,0 +1,536 @@
...
+
+  const subGraphName = unique('ka-lifecycle-cli-sg').replace(/[^a-zA-Z0-9-]/g, '');
+  const sub = await runDkgCli(
+    node,
+    ['context-graph', 'create-sub-graph', contextGraphId, subGraphName],
+    60_000,
+  );
+  expectCliOk(sub, 'context-graph create-sub-graph');
+
+  suite = { node, contextGraphId, subGraphName };
```
Comment (otReviewAgent) [inline devnet/ka-lifecycle-cli/automated.test.ts:345]: 🟡 Issue: The new lifecycle suite couples unrelated workflows through one shared fixture

**What's wrong**
The file is still under the hard 1k-line threshold, but its structure already has several concerns packed together: suite setup, fixture writing, CLI parsing, publisher-job polling, WM/SWM/VM assertions, and legacy-route guards. The module-level suite object makes those concerns share mutable devnet state, so the file will become harder to reason about as KA lifecycle coverage grows.

**Example**
The one-shot publish test leaves publisher jobs and VM/SWM data in the shared context graph before the stepwise and alias tests run. Today unique KA names keep this manageable, but future assertions over publisher jobs, graph contents, or route state will have to remember which prior tests polluted the same fixture.

**Suggested direction**
Create a fresh context graph per workflow, or split the suite into smaller workflow-focused modules backed by explicit fixture factories. That keeps the command flows independent and makes later additions less likely to depend on ordering or leftover graph state.

**For Agents**
Refactor devnet/ka-lifecycle-cli/automated.test.ts so each workflow gets its own lifecycle context, or split setup into focused helpers such as createLifecycleContext(), createSharedKa(), waitForPublisherJobFinalized(), and memoryAssertions. Preserve the four tested CLI workflows, but remove the hidden cross-test fixture coupling.

File: devnet/_bootstrap/harness.ts (hunk 8 above)
Comment (Jurij89) [reply devnet/_bootstrap/harness.ts:102]: Validated. I’m treating this as non-blocking for this devnet-test PR. The shared parser is intentionally scoped to the devnet harness and reflects the current CLI/test boundary where process output can include non-JSON noise around the final JSON payload. Tightening `--json` into a product-wide stdout-only contract would be a broader CLI behavior change, not a prerequisite for adding this coverage. We already added termination coverage for the parser failure mode and the scenario tests still assert the command-specific response fields that matter for lifecycle correctness.

File: devnet/ka-lifecycle-cli/automated.test.ts (hunk 9 above)
Comment (Jurij89) [reply devnet/ka-lifecycle-cli/automated.test.ts:345]: Validated. I’m also treating this as non-blocking for this PR. The suite uses unique KA names, subgraphs, and job polling to avoid cross-workflow ambiguity, and the current structure keeps the end-to-end lifecycle flow readable enough for the initial devnet coverage. Splitting additional fixtures/modules would be reasonable follow-up hygiene if this suite grows, but it would add churn here without changing the behavioral evidence this PR is meant to provide.
